### PR TITLE
Mfma fp16 wrw

### PR DESCRIPTION
**Pull request by shaojiewang**
Fp16 mfma kernel generator for wrw pass.

### CODE REVIEW

File: driver/conv_driver.cpp [hunk 1]
```diff
@@ -802,55 +818,85 @@ int main(int argc, char **argv) {
         float *device_weight_to_host = NULL;
         if (need_verify) {
             // gen rand
-            gen_rand_vector<float, float>(host_input, static_cast<size_t>(n) * c * hi * wi, 0.0, 1.0);
-            gen_rand_vector<float, float>(host_output, static_cast<size_t>(n) * k * ho * wo, -0.5, 0.5);
-            //gen_rand_vector<float, int>(host_input, static_cast<size_t>(n) * k * hi * wi, -5, 5);
-            //gen_rand_vector<float, int>(host_output, static_cast<size_t>(n) * k * ho * wo, 1, 1);
+            gen_rand_vector<float, float>(host_input, n * c * hi * wi, 0.0, 1.0);
```
Comment (carlushuang) [inline driver/conv_driver.cpp:821]: please dont change static_cast<size_t>

File: driver/conv_driver.cpp [hunk 2]
```diff
@@ -920,6 +980,7 @@ int main(int argc, char **argv) {
             }
             printf("\n");
         }
+        dump_arg(&conv_args);
```
Comment (carlushuang) [inline driver/conv_driver.cpp:983]: dump_arg un-wanted here

File: driver/igemm_wrw_gtc_driver.h [hunk 3]
```diff
@@ -201,11 +150,12 @@ class igemm_wrw_gtc_t {
         int gemm_n_per_block         = tunable->gemm_n_per_block;
         int gemm_k_per_block         = tunable->gemm_k_per_block;
         int gemm_k_global_split      = tunable->gemm_k_global_split;
+        int e_x                      = tunable->nxe == 0 ? 1 : tunable->nxe;
 
         gemm_k_global_split = update_gemm_k_global_split(arg, tunable);
 
         int gemm_m = k / group ;
-        int gemm_n = (c / group) * y * x;
+        int gemm_n = (c / group) * y * (((x + e_x - 1) / e_x) * e_x);
```
Comment (carlushuang) [inline driver/igemm_wrw_gtc_driver.h:158]: does this change affect fp32?

File: igemm/algo/igemm_fwd_gtc.py [hunk 4]
```diff
@@ -76,7 +76,7 @@ def __init__(self, mc, inline = False):
         self.declare_arg("s_c")
 
     def name(self):
-        return '.v_set_flag_hw'
+        return '.v_set_flag_c'
```
Comment (carlushuang) [inline igemm/algo/igemm_fwd_gtc.py:79]: why change name to `v_set_flag_c`?

File: igemm/igemm_host_driver.py [hunk 5]
```diff
@@ -72,3 +72,19 @@ def get_dict_with_default(some_dict, key, default_value):
         rtn = hip_builder.compile()
         if not rtn:
             assert False
+
+    # compile reduction code
+    hip_src = os.path.join(cpp_dir, "device_reduction", "gpu_reduction_fp16.cpp")
```
Comment (carlushuang) [inline igemm/igemm_host_driver.py:77]: this is for wrw only and seems not so general. we need think about a way to generalize these approach.

**Review by carlushuang — COMMENTED** [review]
some file mode also changed, which is not preferred here. Also, we'd better make changes more generic and less hacky

File: driver/conv_driver.cpp (hunk 1 above)
Comment (shaojiewang) [reply driver/conv_driver.cpp:821]: OK

File: driver/conv_driver.cpp (hunk 2 above)
Comment (shaojiewang) [reply driver/conv_driver.cpp:983]: OK

File: driver/igemm_wrw_gtc_driver.h (hunk 3 above)
Comment (shaojiewang) [reply driver/igemm_wrw_gtc_driver.h:158]: no, for fp32 cases, `e_x` will always be 1. And fp32 cases also be tested.

File: igemm/algo/igemm_fwd_gtc.py (hunk 4 above)
Comment (shaojiewang) [reply igemm/algo/igemm_fwd_gtc.py:79]: It is actually `v_set_flag_c`. It is used to check if c overflows. Code in master branch made mistake here.

File: igemm/igemm_host_driver.py (hunk 5 above)
Comment (shaojiewang) [reply igemm/igemm_host_driver.py:77]: ok

**Comment by shaojiewang** [comment]
Out of date

**Pull request closed without merging**